apps/aecore/lib/aecore/poi/poi_proof.ex

```diff
@@ -0,0 +1,231 @@
+  @spec get_readonly_proof_trie(PoiProof.t()) :: Trie.t()
+  defp get_readonly_proof_trie(%PoiProof{} = poi_proof) do
+    len = Storage.max_rlp_len()
+    Trie.new(
+      ExternalDB.init(get_proof_readonly_handles(poi_proof)),
+      <<0::size(len)-unit(8)>> #this will avoid writing the initial root hash to the readonly DB
+    )
+  end
+
+  #Invokes proof construction on the Poi proof. Uses the PoiDB wrapper for obtaining imperative behaviour.
+  @spec invoke_proof_construction(PoiProof.t(), Trie.t(), Trie.key()) :: {:ok, Trie.value(), Map.t()} | {:error, :key_not_found}
+  defp invoke_proof_construction(%PoiProof{} = poi_proof, %Trie{} = trie, key) do
+    proof_trie = get_proof_construction_trie(poi_proof)
+    {value, _} = Proof.construct_proof({trie, key, proof_trie})
+    new_proof_db = PoiDB.finalize()
+    case value do
+      nil ->
+        {:error, :key_not_found}
+      _ ->
+        {:ok, value, new_proof_db}
+    end
+  end
+
+  @doc """
+  Adds the value associated with the given key in the given trie to the Poi proof
+  """
+  @spec add_to_poi(PoiProof.t(), Trie.t(), Trie.key()) :: {:ok, Trie.value(), PoiProof.t()} | {:error, :wrong_root_hash | :key_not_found}
+  def add_to_poi(%PoiProof{root_hash: root_hash} = poi_proof, %Trie{} = trie, key) do
+    case trie_root_hash(trie) do
+      ^root_hash ->
+        case invoke_proof_construction(poi_proof, trie, key) do
+          {:error, _} = err ->
+            err
+          {:ok, value, proof_db} ->
+            {:ok, value, %PoiProof{poi_proof | db: proof_db}}
+        end
+      _ ->
+        {:error, :wrong_root_hash}
+    end
+  end
+
+  @doc """
+  Verifies if an entry is present under the given key in the Poi proof
+  """
+  @spec verify_poi_entry(PoiProof.t(), Trie.key, Trie.value()) :: boolean()
+  def verify_poi_entry(%PoiProof{} = poi_proof, key, serialized_value) do
+    root_hash = poi_root_hash_to_trie_root_hash(poi_proof)
+    proof_trie = get_readonly_proof_trie(poi_proof)
+    PatriciaMerkleTree.verify_proof(key, serialized_value, root_hash, proof_trie)
+  end
+
+  @doc """
+  Lookups the entry assiociated with the given key in the Poi proof
+  """
+  @spec lookup_in_poi(PoiProof.t(), Trie.key()) :: {:ok, Trie.value()} | :error
+  def lookup_in_poi(%PoiProof{} = poi_proof, key) do
+    root_hash = poi_root_hash_to_trie_root_hash(poi_proof)
+    proof_trie = get_readonly_proof_trie(poi_proof)
+    PatriciaMerkleTree.lookup_proof(key, root_hash, proof_trie)
+  end
+
+  @doc """
+  Serializes the poi proof to a list
+  """
+  @spec encode_to_list(PoiProof.t()) :: list()
+  def encode_to_list(%PoiProof{root_hash: :empty}) do
+    []
+  end
+
+  def encode_to_list(%PoiProof{root_hash: root_hash, db: db}) do
+    contents =
+      db
+      |> Enum.map(fn {key, value} -> [key, ExRLP.decode(value)] end)
+      |> Enum.sort_by(fn [key, _] -> key end) #the specification requires keys to be sorted in serialized POI's
+    [[root_hash, contents]]
```

Comment:
Yes, this is the serialization used by epoch 0.16.
I discovered this when I noticed that the aecore_poi_epoch_compatibility_and_functional_test was failing because my code was not able to produce the same serialization as epoch. After I inspected the rlp decoded data it turned out that the proofs in epoch are enclosed in a double list.
I wont resolve this conversation here as I'm not sure why epoch uses a double list here.